File: java/ql/src/Security/CWE/CWE-190/ComparisonWithWiderType.java
class Test {
	public static void main(String[] args) {
		
		{		
			int BIGNUM = Integer.MAX_VALUE;
			long MAXGET = Short.MAX_VALUE + 1;
			
			char[] buf = new char[BIGNUM];

			short bytesReceived = 0;
			
			// BAD: 'bytesReceived' is compared with a value of wider type.
			// 'bytesReceived' overflows before reaching MAXGET,
			// causing an infinite loop.
			while (bytesReceived < MAXGET) {
				bytesReceived += getFromInput(buf, bytesReceived);
			}
		}
		
		{
			long bytesReceived2 = 0;
			
			// GOOD: 'bytesReceived2' has a type at least as wide as MAXGET.
			while (bytesReceived2 < MAXGET) {
				bytesReceived2 += getFromInput(buf, bytesReceived2);
			}
		}
		
	}
	
	public static int getFromInput(char[] buf, short pos) {
		// write to buf
		// ...
		return 1;
	}
}

File: cpp/ql/src/Security/CWE/CWE-190/ComparisonWithWiderType.c
void main(int argc, char **argv) {
	uint32_t big_num = INT32_MAX;
	char buf[big_num];
	int16_t bytes_received = 0;
	int max_get = INT16_MAX + 1;

	// BAD: 'bytes_received' is compared with a value of a wider type.
	// 'bytes_received' overflows before  reaching 'max_get',
	// causing an infinite loop
	while (bytes_received < max_get)
		bytes_received += get_from_input(buf, bytes_received);
	}

	uint32_t bytes_received = 0;

	// GOOD: 'bytes_received2' has a type  at least as wide as 'max_get'
	while (bytes_received < max_get) {
		bytes_received += get_from_input(buf, bytes_received);
	}

}


int getFromInput(char *buf, short pos) {
	// write to buf
	// ...
	return 1;
}
